### PR TITLE
Optimize KDA recompute backward

### PR DESCRIPTION
**Pull request by drisspg**
Stacked PRs:
 * #369
 * #368
 * __->__#367

--- --- ---

Optimize KDA recompute backward

## Agent note

Reduce the recompute backward boundary by removing unnecessary fills, using the persistent tape
engine, fusing dense dV work into delta-H, and pinning the measured W/U and dAv configurations.

## Performance

The primary recompute backward improves from about 3.07 ms to 2.72 ms on B200.

## Test Plan

```bash
pytest -n 6 -q test/kda/mega
prek
```

### CODE REVIEW

File: attn_gym/linear/kda/ops.py [hunk 1]
```diff
@@ -57,6 +56,22 @@
...
 
+_CHUNK_BWD_RECOMPUTE_ARGS = (
+    "(Tensor q, Tensor k, Tensor v, Tensor cumulative_gate, Tensor beta, "
+    "Tensor? cu_seqlens, Tensor? chunk_offsets, Tensor? d_output, "
+    "Tensor? d_final_state, {initial_state}, bool fastmath, bool autotune)"
+)
+torch.library.define(
+    "attn_gym::kda_chunk_bwd_recompute_tape",
+    _CHUNK_BWD_RECOMPUTE_ARGS.format(initial_state="Tensor? initial_state")
+    + " -> (Tensor, Tensor, Tensor, Tensor, Tensor)",
```
[the remark this answers is not in the document]
Comment (drisspg) [reply attn_gym/linear/kda/ops.py:71]: The recompute-factor operators are consumed by the Mega training backend in PR #369. The legacy fused backend intentionally retains saved factors for its faster speed-oriented mode, while Mega owns the local-recompute memory tradeoff.

File: attn_gym/linear/kda/fwd/cute/chunk_kda_fwd_intra_engine.py [hunk 2]
```diff
@@ -0,0 +1,862 @@
...
+        capacity = tokens // chunk_size
+        cu_seqlens = chunk_offsets = None
+    else:
+        metadata.validate_chunk_size(chunk_size)
+        capacity = metadata.capacity
+        cu_seqlens = metadata.cu_seqlens
+        chunk_offsets = metadata.chunk_offsets
+    aqk = torch.empty((1, tokens, heads, chunk_size), device=q.device, dtype=torch.bfloat16)
+    offdiag_blocks = 6
+    akkod_rows = max(capacity * offdiag_blocks, 1)
```
[the remark this answers is not in the document]
Comment (drisspg) [reply attn_gym/linear/kda/fwd/cute/chunk_kda_fwd_intra_engine.py]: Fixed: the engine now preserves the exact zero-capacity AkkOD shape and returns before launch. Dense and all-empty packed zero-capacity shapes are covered by the Mega intra-engine tests.

File: attn_gym/linear/kda/fwd/cute/chunk_kda_fwd_intra_engine.py [hunk 3]
```diff
@@ -0,0 +1,862 @@
...
+    # Dense execution writes every row. Ragged capacity slack stays zero for downstream consumers.
+    akkod_factory = torch.empty if metadata is None else torch.zeros
+    akkod = akkod_factory((akkod_rows, heads * 256), device=q.device, dtype=torch.float32)
+    akkd = torch.empty((1, tokens, heads, 16), device=q.device, dtype=torch.float32)
+    compiled = _compile_intra_engine_fwd(heads, head_dim, metadata is not None)
```
[the remark this answers is not in the document]
Comment (drisspg) [reply attn_gym/linear/kda/fwd/cute/chunk_kda_fwd_intra_engine.py]: Fixed: requires_int64_abi now participates in the compile-cache key, the wide ABI uses cute.sym_int64, and address arithmetic widens before multiplication. Forced-int64 forward/backward equivalence is covered in test_kda_int64_offsets.py.

File: attn_gym/linear/kda/bwd/cute/chunk_kda_bwd_intra.py [hunk 4]
```diff
@@ -2189,12 +2205,20 @@ def launch(
...
         )
-        return (
-            args.dq2,
-            args.dk2,
-            args.dg2,
-            args.db + args.db_partial.sum(0).unsqueeze(0),
+        db2 = torch.empty_like(args.db)
+        elements = args.db.numel()
+        block = 256
+        _finalize_db_kernel[(triton.cdiv(elements, block),)](
```
[the remark this answers is not in the document]
Comment (drisspg) [reply attn_gym/linear/kda/bwd/cute/chunk_kda_bwd_intra.py:2224]: Fixed: dense and packed zero-token inputs return before the beta finalizer launch, preserving the empty db result. The zero-token backward test covers both metadata modes.

File: attn_gym/linear/kda/fwd/triton/chunk_delta_h.py [hunk 5]
```diff
@@ -420,7 +356,7 @@ def _delta_h_launch(
         TensorDescriptor.from_tensor(u, [1, _CHUNK_SIZE, 1, block_value_dim]),
         TensorDescriptor.from_tensor(v_new, [1, _CHUNK_SIZE, 1, block_value_dim]),
         TensorDescriptor.from_tensor(h, [1, 1, 1, key_dim, block_value_dim]),
-        TensorDescriptor.from_tensor(decay, [1, 1, key_dim]),
+        TensorDescriptor.from_tensor(gk, [1, 1, 1, key_dim]),
```
[the remark this answers is not in the document]
Comment (drisspg) [reply attn_gym/linear/kda/fwd/triton/chunk_delta_h.py:361]: Fixed: the eager custom-op implementation clones gk only when it cannot back a TMA descriptor. A misaligned compact-gate regression now passes in eager and torch.compile fullgraph execution.

File: attn_gym/linear/kda/fwd/cute/chunk_kda_fwd.py [hunk 6]
```diff
@@ -316,11 +310,58 @@ def _chunk_kda_bwd_shared(
             chunk_capacity(q.shape[1], cu_seqlens.shape[0] - 1, _CHUNK_SIZE),
             _CHUNK_SIZE,
         )
-
     q, k, v = (_normalize_qkv_layout(tensor) for tensor in (q, k, v))
     _validate_private_abi(q, k, v, cumulative_gate, beta, initial_state)
-    if d_output is None:
-        d_output = v.new_zeros(v.shape)
+    d_output = v.new_zeros(v.shape) if d_output is None else _normalize_qkv_layout(d_output)
+    d_final_state = None if d_final_state is None else _normalize_qkv_layout(d_final_state.float())
```
[the remark this answers is not in the document]
Comment (drisspg) [reply attn_gym/linear/kda/fwd/cute/chunk_kda_fwd.py]: Agreed and fixed in 3636a9d. Packed d_output and d_final_state now use a dedicated compact, 128-byte-aligned normalization: outer-strided and misaligned cotangents are cloned, while already compact/aligned training cotangents remain zero-copy. Added regression coverage for both cotangents and both layouts; the packed public-backward suite passes.

File: attn_gym/linear/kda/fwd/cute/chunk_kda_fwd_intra_engine.py [hunk 7]
```diff
@@ -0,0 +1,1074 @@
...
+    akkod = akkod_factory(
+        (capacity * offdiag_blocks, heads * 256), device=q.device, dtype=torch.float32
+    )
+    akkd = torch.empty((1, tokens, heads, 16), device=q.device, dtype=torch.float32)
+    if capacity == 0:
+        return aqk, akkod, akkd
+
+    q_flat = q[0]
+    k_flat = k[0]
+    g_flat = g[0].contiguous()
```
[the remark this answers is not in the document]
Comment (drisspg) [reply attn_gym/linear/kda/fwd/cute/chunk_kda_fwd_intra_engine.py]: Fixed in 6c023cc. The intra engine now clones compact gate tensors unless they satisfy the 128-byte alignment promised by the TVM-FFI signature; beta receives the analogous 8-byte check. Added a regression with contiguous misaligned gate and beta storage. The normal aligned training path remains zero-copy.

File: attn_gym/linear/kda/fwd/cute/chunk_kda_fwd_intra_engine.py [hunk 8]
```diff
@@ -0,0 +1,1074 @@
...
+        assert tokens % chunk_size == 0
+        capacity = tokens // chunk_size
+        cu_seqlens = chunk_offsets = None
+        num_sequences = 0
+    else:
+        capacity = metadata.capacity
+        cu_seqlens = metadata.cu_seqlens
+        chunk_offsets = metadata.chunk_offsets
+        num_sequences = cu_seqlens.shape[0] - 1
+    _diag_neumann_inverse_kernel[(capacity * blocks_per_chunk, heads)](
```
[the remark this answers is not in the document]
Comment (drisspg) [reply attn_gym/linear/kda/fwd/cute/chunk_kda_fwd_intra_engine.py:108]: This is already guarded at the sole call site: kda_intra_engine_fwd returns (aqk, akkod, akkd) immediately when capacity == 0, before compiling the engine or calling _kda_diag_neumann_inverse. Repo-wide search confirms the inverse helper has no other caller. The dense-zero and all-empty-packed engine tests execute this path and pass without a launch, so no code change is needed.
